{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f16d5639",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from datetime import datetime\n",
    "from typing import Iterable\n",
    "from typing import List, Optional\n",
    "import matplotlib.pyplot as plt  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8882e70a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variety_ID</th>\n",
       "      <th>grain.yield</th>\n",
       "      <th>grain.number</th>\n",
       "      <th>seed.size</th>\n",
       "      <th>anthesis</th>\n",
       "      <th>silking</th>\n",
       "      <th>plant.height</th>\n",
       "      <th>tassel.height</th>\n",
       "      <th>ear.height</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11430</td>\n",
       "      <td>6.714175</td>\n",
       "      <td>2524.895871</td>\n",
       "      <td>300.294106</td>\n",
       "      <td>65.329162</td>\n",
       "      <td>68.088152</td>\n",
       "      <td>214.287692</td>\n",
       "      <td>244.776356</td>\n",
       "      <td>108.314869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3</td>\n",
       "      <td>6.283445</td>\n",
       "      <td>2687.746107</td>\n",
       "      <td>267.936893</td>\n",
       "      <td>66.266127</td>\n",
       "      <td>72.116642</td>\n",
       "      <td>221.942849</td>\n",
       "      <td>246.568521</td>\n",
       "      <td>112.339860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A310</td>\n",
       "      <td>5.648845</td>\n",
       "      <td>2098.978720</td>\n",
       "      <td>304.175569</td>\n",
       "      <td>66.237934</td>\n",
       "      <td>69.250734</td>\n",
       "      <td>221.825845</td>\n",
       "      <td>240.521063</td>\n",
       "      <td>124.086819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A347</td>\n",
       "      <td>5.493820</td>\n",
       "      <td>2119.242474</td>\n",
       "      <td>291.957786</td>\n",
       "      <td>67.707086</td>\n",
       "      <td>70.470427</td>\n",
       "      <td>209.095458</td>\n",
       "      <td>243.122323</td>\n",
       "      <td>114.835957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A374</td>\n",
       "      <td>7.505574</td>\n",
       "      <td>3357.301585</td>\n",
       "      <td>258.256427</td>\n",
       "      <td>67.818736</td>\n",
       "      <td>70.948375</td>\n",
       "      <td>218.698006</td>\n",
       "      <td>245.216113</td>\n",
       "      <td>114.789468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>W604S</td>\n",
       "      <td>6.493161</td>\n",
       "      <td>2465.328005</td>\n",
       "      <td>302.815798</td>\n",
       "      <td>68.165436</td>\n",
       "      <td>71.256180</td>\n",
       "      <td>218.385125</td>\n",
       "      <td>249.157640</td>\n",
       "      <td>116.467945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>W64A</td>\n",
       "      <td>6.386390</td>\n",
       "      <td>2560.871368</td>\n",
       "      <td>294.354638</td>\n",
       "      <td>65.585955</td>\n",
       "      <td>69.247288</td>\n",
       "      <td>208.377146</td>\n",
       "      <td>237.077976</td>\n",
       "      <td>107.666317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>W9</td>\n",
       "      <td>5.061490</td>\n",
       "      <td>2346.467710</td>\n",
       "      <td>246.163726</td>\n",
       "      <td>63.539328</td>\n",
       "      <td>68.091173</td>\n",
       "      <td>210.127044</td>\n",
       "      <td>232.934226</td>\n",
       "      <td>106.582606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>W95115</td>\n",
       "      <td>5.621730</td>\n",
       "      <td>2142.086821</td>\n",
       "      <td>292.216995</td>\n",
       "      <td>65.304476</td>\n",
       "      <td>69.832943</td>\n",
       "      <td>208.493574</td>\n",
       "      <td>229.963484</td>\n",
       "      <td>105.211562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Wf9</td>\n",
       "      <td>6.272347</td>\n",
       "      <td>2454.592094</td>\n",
       "      <td>299.681149</td>\n",
       "      <td>66.779881</td>\n",
       "      <td>71.883454</td>\n",
       "      <td>219.639271</td>\n",
       "      <td>246.862014</td>\n",
       "      <td>113.070372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variety_ID  grain.yield  grain.number   seed.size   anthesis    silking  \\\n",
       "0        11430     6.714175   2524.895871  300.294106  65.329162  68.088152   \n",
       "1           A3     6.283445   2687.746107  267.936893  66.266127  72.116642   \n",
       "2         A310     5.648845   2098.978720  304.175569  66.237934  69.250734   \n",
       "3         A347     5.493820   2119.242474  291.957786  67.707086  70.470427   \n",
       "4         A374     7.505574   3357.301585  258.256427  67.818736  70.948375   \n",
       "..         ...          ...           ...         ...        ...        ...   \n",
       "241      W604S     6.493161   2465.328005  302.815798  68.165436  71.256180   \n",
       "242       W64A     6.386390   2560.871368  294.354638  65.585955  69.247288   \n",
       "243         W9     5.061490   2346.467710  246.163726  63.539328  68.091173   \n",
       "244     W95115     5.621730   2142.086821  292.216995  65.304476  69.832943   \n",
       "245        Wf9     6.272347   2454.592094  299.681149  66.779881  71.883454   \n",
       "\n",
       "     plant.height  tassel.height  ear.height  \n",
       "0      214.287692     244.776356  108.314869  \n",
       "1      221.942849     246.568521  112.339860  \n",
       "2      221.825845     240.521063  124.086819  \n",
       "3      209.095458     243.122323  114.835957  \n",
       "4      218.698006     245.216113  114.789468  \n",
       "..            ...            ...         ...  \n",
       "241    218.385125     249.157640  116.467945  \n",
       "242    208.377146     237.077976  107.666317  \n",
       "243    210.127044     232.934226  106.582606  \n",
       "244    208.493574     229.963484  105.211562  \n",
       "245    219.639271     246.862014  113.070372  \n",
       "\n",
       "[246 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traits = pd.read_csv('./maize/maize_mean_pheno.csv')\n",
    "traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c62be0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>SYN83</th>\n",
       "      <th>PZE.101000060</th>\n",
       "      <th>PZE.101000088</th>\n",
       "      <th>PZE.101000083</th>\n",
       "      <th>PZE.101000108</th>\n",
       "      <th>PZE.101000111</th>\n",
       "      <th>PZE.101000161</th>\n",
       "      <th>PZE.101000169</th>\n",
       "      <th>PZE.101000209</th>\n",
       "      <th>...</th>\n",
       "      <th>PZE.110111381</th>\n",
       "      <th>PZE.110111385</th>\n",
       "      <th>PZE.110111388</th>\n",
       "      <th>PZE.110111408</th>\n",
       "      <th>PZE.110111412</th>\n",
       "      <th>PZE.110111417</th>\n",
       "      <th>PZE.110111438</th>\n",
       "      <th>PZE.110111440</th>\n",
       "      <th>PZE.110111453</th>\n",
       "      <th>PZE.110111485</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11430</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A310</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A347</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A374</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>W604S</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>W64A</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>W9</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>W95115</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Wf9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 41723 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    individual  SYN83  PZE.101000060  PZE.101000088  PZE.101000083  \\\n",
       "0        11430      0              2              2              0   \n",
       "1           A3      0              2              2              0   \n",
       "2         A310      2              2              2              2   \n",
       "3         A347      2              0              2              2   \n",
       "4         A374      2              0              2              2   \n",
       "..         ...    ...            ...            ...            ...   \n",
       "241      W604S      0              2              2              0   \n",
       "242       W64A      0              2              2              0   \n",
       "243         W9      2              2              2              0   \n",
       "244     W95115      0              2              2              0   \n",
       "245        Wf9      0              2              2              0   \n",
       "\n",
       "     PZE.101000108  PZE.101000111  PZE.101000161  PZE.101000169  \\\n",
       "0                0              2              2              0   \n",
       "1                0              2              2              0   \n",
       "2                2              2              2              0   \n",
       "3                2              2              1              0   \n",
       "4                2              2              0              2   \n",
       "..             ...            ...            ...            ...   \n",
       "241              2              0              2              0   \n",
       "242              0              2              2              0   \n",
       "243              2              2              2              2   \n",
       "244              2              2              2              0   \n",
       "245              0              2              2              0   \n",
       "\n",
       "     PZE.101000209  ...  PZE.110111381  PZE.110111385  PZE.110111388  \\\n",
       "0                0  ...              2              2              2   \n",
       "1                0  ...              0              0              0   \n",
       "2                0  ...              2              2              2   \n",
       "3                0  ...              2              2              2   \n",
       "4                2  ...              2              2              2   \n",
       "..             ...  ...            ...            ...            ...   \n",
       "241              2  ...              2              2              2   \n",
       "242              0  ...              0              0              0   \n",
       "243              2  ...              2              2              2   \n",
       "244              2  ...              1              2              1   \n",
       "245              0  ...              0              0              0   \n",
       "\n",
       "     PZE.110111408  PZE.110111412  PZE.110111417  PZE.110111438  \\\n",
       "0                2              2              2              2   \n",
       "1                0              0              0              0   \n",
       "2                2              2              2              2   \n",
       "3                2              2              2              2   \n",
       "4                2              2              2              2   \n",
       "..             ...            ...            ...            ...   \n",
       "241              2              2              2              2   \n",
       "242              0              0              0              0   \n",
       "243              2              2              2              2   \n",
       "244              1              1              1              1   \n",
       "245              0              0              0              0   \n",
       "\n",
       "     PZE.110111440  PZE.110111453  PZE.110111485  \n",
       "0                2              2              2  \n",
       "1                0              0              0  \n",
       "2                2              2              2  \n",
       "3                2              2              2  \n",
       "4                2              2              2  \n",
       "..             ...            ...            ...  \n",
       "241              2              2              2  \n",
       "242              0              0              0  \n",
       "243              2              2              2  \n",
       "244              1              1              1  \n",
       "245              0              0              0  \n",
       "\n",
       "[246 rows x 41723 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers = pd.read_csv('./maize/maize_markers.csv')\n",
    "markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0c0ded9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snp_heterozygosity(df, count_nans_as_invalid=True):\n",
    "    \"\"\"\n",
    "    Computes observed and expected heterozygosity per column and the inbreeding\n",
    "    coefficient F.\n",
    "\n",
    "    Accepts SNP genotypes encoded as:\n",
    "      0 -> homozygote REF/REF\n",
    "      1 -> heterozygote REF/ALT\n",
    "      2 -> homozygote ALT/ALT\n",
    "\n",
    "    PARAMETERS:\n",
    "      df: DataFrame with the data\n",
    "      count_nans_as_invalid: if True, NaNs are excluded from the denominator\n",
    "\n",
    "    RETURNS:\n",
    "      DataFrame with Ho and He per locus and average.\n",
    "\n",
    "    It also creates:\n",
    "      - global variable F\n",
    "      - global variable heterozygosity_table\n",
    "    \"\"\"\n",
    "    \n",
    "    # Automatically use the first column's name as the column to ignore\n",
    "    ignore_cols = (df.columns[0],)\n",
    "\n",
    "    cols = [c for c in df.columns if c not in ignore_cols]\n",
    "\n",
    "    observed = {}\n",
    "    expected = {}\n",
    "\n",
    "    for col in cols:\n",
    "        series_raw = df[col]\n",
    "\n",
    "        # Denominator based on parameter count_nans_as_invalid\n",
    "        if count_nans_as_invalid:\n",
    "            # Exclude NaNs from the denominator\n",
    "            series_valid = series_raw.dropna()\n",
    "            total = len(series_valid)\n",
    "        else:\n",
    "            # Include NaNs in the denominator\n",
    "            series_valid = series_raw.dropna()\n",
    "            total = len(series_raw)\n",
    "\n",
    "        if total == 0:\n",
    "            observed[col] = np.nan\n",
    "            expected[col] = np.nan\n",
    "            continue\n",
    "\n",
    "        # Convert entries to 0/1/2 if valid\n",
    "        def to_int_if_valid(x):\n",
    "            try:\n",
    "                if pd.isna(x):\n",
    "                    return np.nan\n",
    "                xi = int(x)\n",
    "                if xi in (0, 1, 2):\n",
    "                    return xi\n",
    "                return np.nan\n",
    "            except Exception:\n",
    "                return np.nan\n",
    "\n",
    "        genotypes_valid = series_valid.map(to_int_if_valid).dropna().astype(int)\n",
    "\n",
    "        # Observed heterozygosity (Ho): proportion of heterozygotes (1)\n",
    "        n_het = (genotypes_valid == 1).sum()\n",
    "        Ho = n_het / total if total > 0 else np.nan\n",
    "        observed[col] = Ho\n",
    "\n",
    "        # Expected heterozygosity (He): 2*p*q\n",
    "        n0 = (genotypes_valid == 0).sum()  # homozygote REF/REF\n",
    "        n1 = (genotypes_valid == 1).sum()  # heterozygote REF/ALT\n",
    "        n2 = (genotypes_valid == 2).sum()  # homozygote ALT/ALT\n",
    "        n_valid = n0 + n1 + n2  # number of valid genotypes\n",
    "\n",
    "        if n_valid == 0:\n",
    "            He = np.nan\n",
    "        else:\n",
    "            # Allele counts: REF allele = 2*n0 + n1 ; ALT allele = 2*n2 + n1\n",
    "            ref_count = 2 * n0 + n1\n",
    "            alt_count = 2 * n2 + n1\n",
    "            denom = 2 * n_valid  # Total number of alleles\n",
    "            p = ref_count / denom\n",
    "            q = alt_count / denom\n",
    "            He = 1 - (p**2 + q**2)  # equivalent to 2*p*q\n",
    "\n",
    "        expected[col] = He\n",
    "\n",
    "    # Build final result table\n",
    "    result = pd.DataFrame(\n",
    "        [observed, expected],\n",
    "        index=[\"observed_heterozygosity\", \"expected_heterozygosity\"]\n",
    "    )\n",
    "    result[\"average\"] = result.mean(axis=1)\n",
    "\n",
    "    # Compute global averages and F\n",
    "    Ho_avg = result.loc[\"observed_heterozygosity\", cols].mean()\n",
    "    He_avg = result.loc[\"expected_heterozygosity\", cols].mean()\n",
    "    F_value = 1 - (Ho_avg / He_avg) if (He_avg is not None and not np.isnan(He_avg) and He_avg != 0) else np.nan\n",
    "\n",
    "    # Create global variable F\n",
    "    globals()['F'] = F_value\n",
    "\n",
    "    # Create global variable heterozygosity_table\n",
    "    globals()['heterozygosity_table'] = result\n",
    "\n",
    "    print(\"Inbreeding coefficient F:\", F_value)\n",
    "    print(\"\")\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "179b742d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inbreeding coefficient F: 0.9740991551394081\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SYN83</th>\n",
       "      <th>PZE.101000060</th>\n",
       "      <th>PZE.101000088</th>\n",
       "      <th>PZE.101000083</th>\n",
       "      <th>PZE.101000108</th>\n",
       "      <th>PZE.101000111</th>\n",
       "      <th>PZE.101000161</th>\n",
       "      <th>PZE.101000169</th>\n",
       "      <th>PZE.101000209</th>\n",
       "      <th>PZE.101000256</th>\n",
       "      <th>...</th>\n",
       "      <th>PZE.110111385</th>\n",
       "      <th>PZE.110111388</th>\n",
       "      <th>PZE.110111408</th>\n",
       "      <th>PZE.110111412</th>\n",
       "      <th>PZE.110111417</th>\n",
       "      <th>PZE.110111438</th>\n",
       "      <th>PZE.110111440</th>\n",
       "      <th>PZE.110111453</th>\n",
       "      <th>PZE.110111485</th>\n",
       "      <th>average</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>observed_heterozygosity</th>\n",
       "      <td>0.077236</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.004065</td>\n",
       "      <td>0.004065</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028455</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.016260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.004065</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.009670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>expected_heterozygosity</th>\n",
       "      <td>0.481749</td>\n",
       "      <td>0.400026</td>\n",
       "      <td>0.269573</td>\n",
       "      <td>0.487433</td>\n",
       "      <td>0.176086</td>\n",
       "      <td>0.288486</td>\n",
       "      <td>0.382998</td>\n",
       "      <td>0.484004</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.417377</td>\n",
       "      <td>...</td>\n",
       "      <td>0.314099</td>\n",
       "      <td>0.319023</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.296343</td>\n",
       "      <td>0.304052</td>\n",
       "      <td>0.298929</td>\n",
       "      <td>0.373346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 41723 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            SYN83  PZE.101000060  PZE.101000088  \\\n",
       "observed_heterozygosity  0.077236       0.024390       0.004065   \n",
       "expected_heterozygosity  0.481749       0.400026       0.269573   \n",
       "\n",
       "                         PZE.101000083  PZE.101000108  PZE.101000111  \\\n",
       "observed_heterozygosity       0.004065       0.008130       0.000000   \n",
       "expected_heterozygosity       0.487433       0.176086       0.288486   \n",
       "\n",
       "                         PZE.101000161  PZE.101000169  PZE.101000209  \\\n",
       "observed_heterozygosity       0.028455       0.008130       0.008130   \n",
       "expected_heterozygosity       0.382998       0.484004       0.298929   \n",
       "\n",
       "                         PZE.101000256  ...  PZE.110111385  PZE.110111388  \\\n",
       "observed_heterozygosity       0.016260  ...       0.008130       0.008130   \n",
       "expected_heterozygosity       0.417377  ...       0.314099       0.319023   \n",
       "\n",
       "                         PZE.110111408  PZE.110111412  PZE.110111417  \\\n",
       "observed_heterozygosity       0.008130       0.008130       0.008130   \n",
       "expected_heterozygosity       0.298929       0.298929       0.298929   \n",
       "\n",
       "                         PZE.110111438  PZE.110111440  PZE.110111453  \\\n",
       "observed_heterozygosity       0.008130       0.004065       0.008130   \n",
       "expected_heterozygosity       0.298929       0.296343       0.304052   \n",
       "\n",
       "                         PZE.110111485   average  \n",
       "observed_heterozygosity       0.008130  0.009670  \n",
       "expected_heterozygosity       0.298929  0.373346  \n",
       "\n",
       "[2 rows x 41723 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_heterozygosity(markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "be97eabb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snp_pairwise_relatedness(df=None, decimals: int = 4) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Compute pairwise relatedness as Pearson correlation between rows (individuals).\n",
    "\n",
    "    Notes:\n",
    "      - First column is treated as the sample ID.\n",
    "      - The remaining columns are SNPs coded as 0/1/2.\n",
    "      - Uses pairwise-complete Pearson correlation.\n",
    "      - Produces global variables r and relatedness_table.\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"Step 1/6: Checking input dataframe...\")\n",
    "    global df_gen\n",
    "    if df is None:\n",
    "        df = df_gen\n",
    "\n",
    "    if df is None:\n",
    "        raise ValueError(\"No dataframe provided and global df_gen is not defined.\")\n",
    "\n",
    "    nrows = len(df)\n",
    "    if nrows < 2:\n",
    "        print(\"Not enough rows to compute pairwise relatedness.\")\n",
    "        r = pd.DataFrame(columns=[\"pair\", \"relatedness\"])\n",
    "        globals()['r'] = r\n",
    "        globals()['relatedness_table'] = r\n",
    "        return r\n",
    "\n",
    "    print(\"Step 2/6: Extracting individual IDs (first column)...\")\n",
    "    id_col = df.columns[0]\n",
    "    ids = df[id_col].astype(str).tolist()\n",
    "\n",
    "    print(\"Step 3/6: Converting SNP data to numeric format...\")\n",
    "    # Convert remaining columns to numeric (0/1/2), invalid -> NaN\n",
    "    data = df.drop(columns=[id_col]).apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "    print(\"Step 4/6: Computing correlation matrix (this may take a while)...\")\n",
    "    # Transpose so columns = individuals\n",
    "    data_T = data.T\n",
    "    data_T.columns = ids\n",
    "\n",
    "    corr_matrix = data_T.corr(method='pearson')\n",
    "\n",
    "    print(\"Step 5/6: Cleaning correlation matrix (replacing NaN with 0.0)...\")\n",
    "    corr_matrix_filled = corr_matrix.fillna(0.0)\n",
    "\n",
    "    print(\"Step 6/6: Building pairwise relatedness output...\")\n",
    "    results = []\n",
    "    cols = corr_matrix_filled.columns.tolist()\n",
    "\n",
    "    for a, b in itertools.combinations(cols, 2):\n",
    "        val = corr_matrix_filled.at[a, b]\n",
    "        results.append({\n",
    "            \"pair\": f\"{a}_{b}\",\n",
    "            \"relatedness\": round(float(val), decimals)\n",
    "        })\n",
    "\n",
    "    r = pd.DataFrame(results)\n",
    "    globals()['r'] = r\n",
    "    globals()['relatedness_table'] = r\n",
    "\n",
    "    print(\"Done! Pairwise relatedness successfully computed.\")\n",
    "    return r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "10642d25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1/6: Checking input dataframe...\n",
      "Step 2/6: Extracting individual IDs (first column)...\n",
      "Step 3/6: Converting SNP data to numeric format...\n",
      "Step 4/6: Computing correlation matrix (this may take a while)...\n",
      "Step 5/6: Cleaning correlation matrix (replacing NaN with 0.0)...\n",
      "Step 6/6: Building pairwise relatedness output...\n",
      "Done! Pairwise relatedness successfully computed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair</th>\n",
       "      <th>relatedness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11430_A3</td>\n",
       "      <td>0.1015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11430_A310</td>\n",
       "      <td>0.0668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11430_A347</td>\n",
       "      <td>0.0751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11430_A374</td>\n",
       "      <td>0.0574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11430_A375</td>\n",
       "      <td>0.0757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30130</th>\n",
       "      <td>W64A_W95115</td>\n",
       "      <td>0.1018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30131</th>\n",
       "      <td>W64A_Wf9</td>\n",
       "      <td>0.5474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30132</th>\n",
       "      <td>W9_W95115</td>\n",
       "      <td>0.1008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30133</th>\n",
       "      <td>W9_Wf9</td>\n",
       "      <td>0.0731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30134</th>\n",
       "      <td>W95115_Wf9</td>\n",
       "      <td>0.0805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30135 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              pair  relatedness\n",
       "0         11430_A3       0.1015\n",
       "1       11430_A310       0.0668\n",
       "2       11430_A347       0.0751\n",
       "3       11430_A374       0.0574\n",
       "4       11430_A375       0.0757\n",
       "...            ...          ...\n",
       "30130  W64A_W95115       0.1018\n",
       "30131     W64A_Wf9       0.5474\n",
       "30132    W9_W95115       0.1008\n",
       "30133       W9_Wf9       0.0731\n",
       "30134   W95115_Wf9       0.0805\n",
       "\n",
       "[30135 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_pairwise_relatedness(markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1a4cec89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def phenotypic_similarity(\n",
    "    df: Optional[pd.DataFrame],\n",
    "    trait_cols: Optional[List[str]] = None,\n",
    "    decimals: int = 4,\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Compute bounded phenotypic similarity in [0,1] for each trait and each pair of rows,\n",
    "    using R_k = max_k - min_k (range) for normalization.\n",
    "\n",
    "    similarity_k(i,j) = (R_k - abs(x_i - x_j)) / R_k\n",
    "\n",
    "    Behavior:\n",
    "    - If df is None, the function will try to use a global DataFrame named `mor`.\n",
    "      If `mor` is not found, a ValueError is raised.\n",
    "    - If trait_cols is None, numeric columns are auto-selected excluding the FIRST\n",
    "      column (treated as the sample/individual ID).\n",
    "    - If R_k == 0 (all values identical) similarity is set to 1 for that trait.\n",
    "    - If either x_i or x_j is NaN → similarity is NaN for that trait.\n",
    "    - Output: DataFrame with \"pair\" column (using values from the FIRST column if present,\n",
    "      otherwise 1-based indices \"i_j\") and one column per trait.\n",
    "    - Final DataFrame is assigned to the global variable `z` and also returned.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame or None\n",
    "        Input dataframe containing phenotypic trait columns. If None, uses global `mor`.\n",
    "    trait_cols : list[str] | None\n",
    "        List of trait column names to use. If None, auto-select numeric columns\n",
    "        (excluding the first column).\n",
    "    decimals : int\n",
    "        Number of decimals to round similarity values.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        DataFrame named `z` with columns: 'pair' and one column per selected trait.\n",
    "    \"\"\"\n",
    "    # fallback to global 'mor' if df not provided\n",
    "    if df is None:\n",
    "        if 'mor' not in globals():\n",
    "            raise ValueError(\n",
    "                \"df not provided and no global DataFrame named 'mor' found. \"\n",
    "                \"Please pass df or create 'mor'.\"\n",
    "            )\n",
    "        df = globals()['mor']\n",
    "\n",
    "    if not isinstance(df, pd.DataFrame):\n",
    "        raise TypeError(\"df must be a pandas DataFrame or None.\")\n",
    "\n",
    "    cols = list(df.columns)\n",
    "    if len(cols) == 0:\n",
    "        raise ValueError(\"Provided DataFrame has no columns.\")\n",
    "\n",
    "    # The FIRST column is treated as the ID column and excluded from trait selection\n",
    "    id_col = cols[0]\n",
    "    data_cols = cols[1:]  # candidate trait columns\n",
    "\n",
    "    # select traits\n",
    "    if trait_cols is not None:\n",
    "        if not isinstance(trait_cols, (list, tuple)):\n",
    "            raise TypeError(\"trait_cols must be a list/tuple of column names or None.\")\n",
    "        missing = [c for c in trait_cols if c not in cols]\n",
    "        if missing:\n",
    "            raise KeyError(f\"The following requested columns do not exist in the DataFrame: {missing}\")\n",
    "        # disallow using the id column as a trait\n",
    "        if id_col in trait_cols:\n",
    "            raise ValueError(f\"The first column ('{id_col}') is treated as ID and cannot be used as a trait.\")\n",
    "        non_numeric = [c for c in trait_cols if not pd.api.types.is_numeric_dtype(df[c])]\n",
    "        if non_numeric:\n",
    "            raise TypeError(f\"The following columns are not numeric: {non_numeric}\")\n",
    "        selected_traits = list(trait_cols)\n",
    "    else:\n",
    "        # auto-select numeric columns excluding the first column\n",
    "        selected_traits = [c for c in data_cols if pd.api.types.is_numeric_dtype(df[c])]\n",
    "\n",
    "    if not selected_traits:\n",
    "        numeric_cols = [c for c in data_cols if pd.api.types.is_numeric_dtype(df[c])]\n",
    "        raise ValueError(\n",
    "            \"No selectable traits found (selected_traits is empty). \"\n",
    "            f\"Numeric columns available (excluding first column '{id_col}'): {numeric_cols}.\"\n",
    "        )\n",
    "\n",
    "    # compute ranges = max - min for the selected traits\n",
    "    ranges = df[selected_traits].max() - df[selected_traits].min()\n",
    "    ranges = ranges[selected_traits]  # ensure same order/index\n",
    "\n",
    "    results = []\n",
    "    nrows = len(df)\n",
    "\n",
    "    # prepare individual names (use the values from the first column; fallback to 1-based indices)\n",
    "    id_values = df.iloc[:, 0].astype(str).tolist()\n",
    "    use_ids = True  # we always have a first column to use as label\n",
    "\n",
    "    for i, j in itertools.combinations(range(nrows), 2):\n",
    "        if use_ids:\n",
    "            ind_i = id_values[i]\n",
    "            ind_j = id_values[j]\n",
    "            pair_label = f\"{ind_i}_{ind_j}\"\n",
    "        else:\n",
    "            pair_label = f\"{i+1}_{j+1}\"\n",
    "\n",
    "        row = {\"pair\": pair_label}\n",
    "        for col in selected_traits:\n",
    "            xi = df.iloc[i][col]\n",
    "            xj = df.iloc[j][col]\n",
    "            R = ranges.loc[col]\n",
    "\n",
    "            if pd.isna(xi) or pd.isna(xj):\n",
    "                sim = float(\"nan\")\n",
    "            else:\n",
    "                diff = abs(xi - xj)\n",
    "                if pd.isna(R) or R == 0:\n",
    "                    sim = 1.0\n",
    "                else:\n",
    "                    sim = (R - diff) / R\n",
    "                    # clip to [0,1]\n",
    "                    if sim < 0:\n",
    "                        sim = 0.0\n",
    "                    elif sim > 1:\n",
    "                        sim = 1.0\n",
    "\n",
    "            row[col] = round(sim, decimals) if pd.notna(sim) else sim\n",
    "\n",
    "        results.append(row)\n",
    "\n",
    "    # create and export global z\n",
    "    global z, phenotypic_similarity_table\n",
    "    z = pd.DataFrame(results)\n",
    "    phenotypic_similarity_table = z\n",
    "    return z\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6e1833d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair</th>\n",
       "      <th>grain.yield</th>\n",
       "      <th>grain.number</th>\n",
       "      <th>seed.size</th>\n",
       "      <th>anthesis</th>\n",
       "      <th>silking</th>\n",
       "      <th>plant.height</th>\n",
       "      <th>tassel.height</th>\n",
       "      <th>ear.height</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11430_A3</td>\n",
       "      <td>0.9207</td>\n",
       "      <td>0.9241</td>\n",
       "      <td>0.7255</td>\n",
       "      <td>0.9132</td>\n",
       "      <td>0.6998</td>\n",
       "      <td>0.8479</td>\n",
       "      <td>0.9627</td>\n",
       "      <td>0.8956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11430_A310</td>\n",
       "      <td>0.8040</td>\n",
       "      <td>0.8015</td>\n",
       "      <td>0.9671</td>\n",
       "      <td>0.9158</td>\n",
       "      <td>0.9134</td>\n",
       "      <td>0.8502</td>\n",
       "      <td>0.9113</td>\n",
       "      <td>0.5909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11430_A347</td>\n",
       "      <td>0.7754</td>\n",
       "      <td>0.8110</td>\n",
       "      <td>0.9293</td>\n",
       "      <td>0.7798</td>\n",
       "      <td>0.8225</td>\n",
       "      <td>0.8968</td>\n",
       "      <td>0.9655</td>\n",
       "      <td>0.8309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11430_A374</td>\n",
       "      <td>0.8544</td>\n",
       "      <td>0.6121</td>\n",
       "      <td>0.6433</td>\n",
       "      <td>0.7695</td>\n",
       "      <td>0.7868</td>\n",
       "      <td>0.9123</td>\n",
       "      <td>0.9908</td>\n",
       "      <td>0.8321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11430_A375</td>\n",
       "      <td>0.9854</td>\n",
       "      <td>0.8945</td>\n",
       "      <td>0.7859</td>\n",
       "      <td>0.9044</td>\n",
       "      <td>0.8835</td>\n",
       "      <td>0.8780</td>\n",
       "      <td>0.9141</td>\n",
       "      <td>0.8528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30130</th>\n",
       "      <td>W64A_W95115</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>0.8048</td>\n",
       "      <td>0.9819</td>\n",
       "      <td>0.9739</td>\n",
       "      <td>0.9564</td>\n",
       "      <td>0.9977</td>\n",
       "      <td>0.8518</td>\n",
       "      <td>0.9363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30131</th>\n",
       "      <td>W64A_Wf9</td>\n",
       "      <td>0.9790</td>\n",
       "      <td>0.9505</td>\n",
       "      <td>0.9548</td>\n",
       "      <td>0.8894</td>\n",
       "      <td>0.8035</td>\n",
       "      <td>0.7762</td>\n",
       "      <td>0.7962</td>\n",
       "      <td>0.8598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30132</th>\n",
       "      <td>W9_W95115</td>\n",
       "      <td>0.8969</td>\n",
       "      <td>0.9048</td>\n",
       "      <td>0.6093</td>\n",
       "      <td>0.8366</td>\n",
       "      <td>0.8702</td>\n",
       "      <td>0.9675</td>\n",
       "      <td>0.9381</td>\n",
       "      <td>0.9644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30133</th>\n",
       "      <td>W9_Wf9</td>\n",
       "      <td>0.7772</td>\n",
       "      <td>0.9496</td>\n",
       "      <td>0.5460</td>\n",
       "      <td>0.6999</td>\n",
       "      <td>0.7174</td>\n",
       "      <td>0.8109</td>\n",
       "      <td>0.7098</td>\n",
       "      <td>0.8317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30134</th>\n",
       "      <td>W95115_Wf9</td>\n",
       "      <td>0.8803</td>\n",
       "      <td>0.8544</td>\n",
       "      <td>0.9367</td>\n",
       "      <td>0.8634</td>\n",
       "      <td>0.8472</td>\n",
       "      <td>0.7785</td>\n",
       "      <td>0.6479</td>\n",
       "      <td>0.7962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30135 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              pair  grain.yield  grain.number  seed.size  anthesis  silking  \\\n",
       "0         11430_A3       0.9207        0.9241     0.7255    0.9132   0.6998   \n",
       "1       11430_A310       0.8040        0.8015     0.9671    0.9158   0.9134   \n",
       "2       11430_A347       0.7754        0.8110     0.9293    0.7798   0.8225   \n",
       "3       11430_A374       0.8544        0.6121     0.6433    0.7695   0.7868   \n",
       "4       11430_A375       0.9854        0.8945     0.7859    0.9044   0.8835   \n",
       "...            ...          ...           ...        ...       ...      ...   \n",
       "30130  W64A_W95115       0.8593        0.8048     0.9819    0.9739   0.9564   \n",
       "30131     W64A_Wf9       0.9790        0.9505     0.9548    0.8894   0.8035   \n",
       "30132    W9_W95115       0.8969        0.9048     0.6093    0.8366   0.8702   \n",
       "30133       W9_Wf9       0.7772        0.9496     0.5460    0.6999   0.7174   \n",
       "30134   W95115_Wf9       0.8803        0.8544     0.9367    0.8634   0.8472   \n",
       "\n",
       "       plant.height  tassel.height  ear.height  \n",
       "0            0.8479         0.9627      0.8956  \n",
       "1            0.8502         0.9113      0.5909  \n",
       "2            0.8968         0.9655      0.8309  \n",
       "3            0.9123         0.9908      0.8321  \n",
       "4            0.8780         0.9141      0.8528  \n",
       "...             ...            ...         ...  \n",
       "30130        0.9977         0.8518      0.9363  \n",
       "30131        0.7762         0.7962      0.8598  \n",
       "30132        0.9675         0.9381      0.9644  \n",
       "30133        0.8109         0.7098      0.8317  \n",
       "30134        0.7785         0.6479      0.7962  \n",
       "\n",
       "[30135 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotypic_similarity(traits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a78d0933",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heritability(\n",
    "    r_df: Optional[pd.DataFrame] = None,\n",
    "    z_df: Optional[pd.DataFrame] = None,\n",
    "    r_col: str = \"relatedness\",\n",
    "    key: str = \"pair\",\n",
    "    F: Optional[float] = None,\n",
    "    ddof: int = 0,\n",
    "    n_bootstrap: int = 1000,\n",
    "    ci: float = 0.95,\n",
    "    random_state: Optional[int] = None\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Computes per-trait heritability with bootstrap standard errors and confidence intervals.\n",
    "\n",
    "    Behavior:\n",
    "      - If r_df is None, tries to use global variable `r`.\n",
    "      - If z_df is None, tries to use global variable `z`.\n",
    "      - If F is None, tries to use global variable `F`.\n",
    "      - Raises ValueError with clear message if any required input is missing.\n",
    "\n",
    "    Returns a DataFrame with columns:\n",
    "      - trait\n",
    "      - heritability (point estimate)\n",
    "      - standard_error (bootstrap SE)\n",
    "      - ci_lower (bootstrap percentile CI)\n",
    "      - ci_upper\n",
    "    Also assigns the returned DataFrame to global `heritability_table`.\n",
    "    \"\"\"\n",
    "\n",
    "    # --- fallback to globals if not provided ---\n",
    "    if r_df is None:\n",
    "        if 'r' in globals():\n",
    "            r_df = globals()['r']\n",
    "        else:\n",
    "            raise ValueError(\"r_df not provided and no global 'r' found. Please pass r_df or create global 'r'.\")\n",
    "\n",
    "    if z_df is None:\n",
    "        if 'z' in globals():\n",
    "            z_df = globals()['z']\n",
    "        else:\n",
    "            raise ValueError(\"z_df not provided and no global 'z' found. Please pass z_df or create global 'z'.\")\n",
    "\n",
    "    if F is None:\n",
    "        if 'F' in globals():\n",
    "            F = globals()['F']\n",
    "        else:\n",
    "            raise ValueError(\"F not provided and no global 'F' found. Please pass F or run heterozygosity() first.\")\n",
    "    F = float(F)\n",
    "\n",
    "    # ------------ Merge ----------------------\n",
    "    merged = pd.merge(r_df[[key, r_col]], z_df, on=key, how=\"inner\")\n",
    "    if merged.empty:\n",
    "        raise ValueError(\"Merge resulted in empty dataframe. Check keys in r_df and z_df.\")\n",
    "\n",
    "    n_pairs = merged.shape[0]\n",
    "\n",
    "    # ------------ Extract and validate r ------\n",
    "    r = pd.to_numeric(merged[r_col], errors=\"coerce\")\n",
    "    if r.isna().any():\n",
    "        raise ValueError(f\"Non-numeric values found in relatedness column '{r_col}'.\")\n",
    "\n",
    "    var_r = r.var(ddof=ddof)\n",
    "    if var_r == 0 or np.isclose(var_r, 0):\n",
    "        raise ValueError(\"Variance of relatedness is zero; cannot compute heritability.\")\n",
    "\n",
    "    # ------------ Trait columns ---------------\n",
    "    trait_cols = [c for c in merged.columns if c not in {key, r_col}]\n",
    "    if not trait_cols:\n",
    "        raise ValueError(\"No trait columns found after merging r_df and z_df.\")\n",
    "\n",
    "    # ------------ Center data -----------------\n",
    "    r_centered = r - r.mean()\n",
    "    z = merged[trait_cols].apply(pd.to_numeric, errors=\"coerce\")\n",
    "    if z.isna().any().any():\n",
    "        raise ValueError(\"Non-numeric values detected in trait columns after conversion.\")\n",
    "    z_centered = z - z.mean()\n",
    "\n",
    "    # ------------ Point estimates --------------\n",
    "    covariances = (r_centered.values.reshape(-1, 1) * z_centered.values).mean(axis=0)\n",
    "    h2_unadjusted = covariances / (2.0 * var_r)\n",
    "    h2_point = h2_unadjusted * (1.0 + F)\n",
    "\n",
    "    # ------------ Bootstrap --------------------\n",
    "    rng = np.random.default_rng(random_state)\n",
    "    B = int(n_bootstrap)\n",
    "    boot_estimates = np.full((B, len(trait_cols)), np.nan)\n",
    "\n",
    "    for b in range(B):\n",
    "        idx = rng.integers(0, n_pairs, size=n_pairs)\n",
    "        r_s = r.values[idx]\n",
    "        z_s = z.values[idx, :]\n",
    "\n",
    "        var_r_s = np.var(r_s, ddof=ddof)\n",
    "        if var_r_s == 0 or np.isclose(var_r_s, 0):\n",
    "            # skip this bootstrap replicate (invalid)\n",
    "            continue\n",
    "\n",
    "        r_s_centered = r_s - r_s.mean()\n",
    "        z_s_centered = z_s - z_s.mean(axis=0)\n",
    "\n",
    "        cov_s = (r_s_centered.reshape(-1, 1) * z_s_centered).mean(axis=0)\n",
    "        h2_unadj_s = cov_s / (2.0 * var_r_s)\n",
    "        h2_s = h2_unadj_s * (1.0 + F)\n",
    "\n",
    "        boot_estimates[b, :] = h2_s\n",
    "\n",
    "    # Remove invalid bootstrap iterations\n",
    "    valid = ~np.all(np.isnan(boot_estimates), axis=1)\n",
    "    boot_valid = boot_estimates[valid]\n",
    "\n",
    "    if boot_valid.shape[0] == 0:\n",
    "        raise RuntimeError(\"All bootstrap replicates invalid (variance of r was zero in all replicates).\")\n",
    "\n",
    "    # ------------ Standard errors + CI ---------\n",
    "    se = np.nanstd(boot_valid, axis=0, ddof=1)\n",
    "\n",
    "    alpha = (1 - ci) / 2\n",
    "    ci_lower = np.nanpercentile(boot_valid, 100 * alpha, axis=0)\n",
    "    ci_upper = np.nanpercentile(boot_valid, 100 * (1 - alpha), axis=0)\n",
    "\n",
    "    # ------------ Final output -----------------\n",
    "    out = pd.DataFrame({\n",
    "        \"trait\": trait_cols,\n",
    "        \"heritability\": h2_point,\n",
    "        \"standard_error\": se,\n",
    "        \"ci_lower\": ci_lower,\n",
    "        \"ci_upper\": ci_upper\n",
    "    })\n",
    "    \n",
    "    out = out.round(4)\n",
    "    \n",
    "    # assign to global convenience variable\n",
    "    globals()['heritability_table'] = out[[\"trait\", \"heritability\", \"standard_error\", \"ci_lower\", \"ci_upper\"]]\n",
    "    print('n_bootstrap:', n_bootstrap)\n",
    "    print('')\n",
    "    print('heritability_table:')\n",
    "    print('')\n",
    "    print(heritability_table)\n",
    "    print('')\n",
    "    print('To access more results, type: F, heterozygosity_table, relatedness_table, phenotypic_similarity_table, heritability_table.')\n",
    "    print('To save all data, type:  save_run()')\n",
    "    return globals()['heritability_table']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7d05b19d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def h2(gen, mor):\n",
    "    snp_heterozygosity(gen)\n",
    "    snp_pairwise_relatedness(gen)\n",
    "    phenotypic_similarity(mor)\n",
    "    heritability()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "079df68a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inbreeding coefficient F: 0.9740991551394081\n",
      "\n",
      "Step 1/6: Checking input dataframe...\n",
      "Step 2/6: Extracting individual IDs (first column)...\n",
      "Step 3/6: Converting SNP data to numeric format...\n",
      "Step 4/6: Computing correlation matrix (this may take a while)...\n",
      "Step 5/6: Cleaning correlation matrix (replacing NaN with 0.0)...\n",
      "Step 6/6: Building pairwise relatedness output...\n",
      "Done! Pairwise relatedness successfully computed.\n",
      "n_bootstrap: 1000\n",
      "\n",
      "heritability_table:\n",
      "\n",
      "           trait  heritability  standard_error  ci_lower  ci_upper\n",
      "0    grain.yield        0.2164          0.0082    0.2009    0.2331\n",
      "1   grain.number        0.1716          0.0079    0.1571    0.1880\n",
      "2      seed.size        0.1145          0.0067    0.1013    0.1273\n",
      "3       anthesis        0.2049          0.0088    0.1888    0.2226\n",
      "4        silking        0.2150          0.0082    0.1990    0.2304\n",
      "5   plant.height        0.0469          0.0078    0.0319    0.0619\n",
      "6  tassel.height        0.0679          0.0091    0.0502    0.0855\n",
      "7     ear.height        0.0535          0.0085    0.0363    0.0695\n",
      "\n",
      "To access more results, type: F, heterozygosity_table, relatedness_table, phenotypic_similarity_table, heritability_table.\n",
      "To save all data, type:  save_run()\n"
     ]
    }
   ],
   "source": [
    "h2(markers,traits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8635a78e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9740991551394081"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8ee6fa15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>heritability</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>ci_lower</th>\n",
       "      <th>ci_upper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>grain.yield</td>\n",
       "      <td>0.2164</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>0.2009</td>\n",
       "      <td>0.2331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>grain.number</td>\n",
       "      <td>0.1716</td>\n",
       "      <td>0.0079</td>\n",
       "      <td>0.1571</td>\n",
       "      <td>0.1880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>seed.size</td>\n",
       "      <td>0.1145</td>\n",
       "      <td>0.0067</td>\n",
       "      <td>0.1013</td>\n",
       "      <td>0.1273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>anthesis</td>\n",
       "      <td>0.2049</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0.1888</td>\n",
       "      <td>0.2226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>silking</td>\n",
       "      <td>0.2150</td>\n",
       "      <td>0.0082</td>\n",
       "      <td>0.1990</td>\n",
       "      <td>0.2304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plant.height</td>\n",
       "      <td>0.0469</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>0.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tassel.height</td>\n",
       "      <td>0.0679</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>0.0502</td>\n",
       "      <td>0.0855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ear.height</td>\n",
       "      <td>0.0535</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>0.0363</td>\n",
       "      <td>0.0695</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           trait  heritability  standard_error  ci_lower  ci_upper\n",
       "0    grain.yield        0.2164          0.0082    0.2009    0.2331\n",
       "1   grain.number        0.1716          0.0079    0.1571    0.1880\n",
       "2      seed.size        0.1145          0.0067    0.1013    0.1273\n",
       "3       anthesis        0.2049          0.0088    0.1888    0.2226\n",
       "4        silking        0.2150          0.0082    0.1990    0.2304\n",
       "5   plant.height        0.0469          0.0078    0.0319    0.0619\n",
       "6  tassel.height        0.0679          0.0091    0.0502    0.0855\n",
       "7     ear.height        0.0535          0.0085    0.0363    0.0695"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "heritability_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b347ff4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_markers_type(df):\n",
    "    global markers_type  # create/update global variable\n",
    "    \n",
    "    marker_cols = df.columns[1:]\n",
    "\n",
    "    # Detect if any marker contains \"/\"\n",
    "    contains_slash = any(\n",
    "        \"/\" in str(v)\n",
    "        for v in df[marker_cols].stack()\n",
    "    )\n",
    "\n",
    "    # Assign marker type\n",
    "    if contains_slash:\n",
    "        markers_type = \"microsatellites\"\n",
    "        print(\"Type of markers detected: Microsatellites\")\n",
    "    else:\n",
    "        markers_type = \"SNPs\"\n",
    "        print(\"Type of markers detected: SNPs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a4988953",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type of markers detected: SNPs\n"
     ]
    }
   ],
   "source": [
    "detect_markers_type(markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ab0f2e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SNPs'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "markers_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a7745d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
